# Exercise 2.4   
Using the variable totalwgt_lb, investigate whether first babies are lighter or heavier than others. 
Compute Cohen’s d to quantify the difference between the groups. 
How does it compare to the difference in pregnancy length?

## Import libraries

In [38]:
import pandas as pd
import nsfg
import thinkstats2
import thinkplot

In [9]:
df = nsfg.ReadFemPreg()
df

caseid  pregordr  howpreg_n  howpreg_p  moscurrp  nowprgdk  pregend1  \
0           1         1        NaN        NaN       NaN       NaN       6.0   
1           1         2        NaN        NaN       NaN       NaN       6.0   
2           2         1        NaN        NaN       NaN       NaN       5.0   
3           2         2        NaN        NaN       NaN       NaN       6.0   
4           2         3        NaN        NaN       NaN       NaN       6.0   
...       ...       ...        ...        ...       ...       ...       ...   
13588   12571         1        NaN        NaN       NaN       NaN       6.0   
13589   12571         2        NaN        NaN       NaN       NaN       3.0   
13590   12571         3        NaN        NaN       NaN       NaN       3.0   
13591   12571         4        NaN        NaN       NaN       NaN       6.0   
13592   12571         5        NaN        NaN       NaN       NaN       6.0   

       pregend2  nbrnaliv  multbrth  ...  laborfor_i  religion_i  metro_i  \
0           NaN       1.0       NaN  ...           0           0        0   
1           NaN       1.0       NaN  ...           0           0        0   
2           NaN       3.0       5.0  ...           0           0        0   
3           NaN       1.0       NaN  ...           0           0        0   
4           NaN       1.0       NaN  ...           0           0        0   
...         ...       ...       ...  ...         ...         ...      ...   
13588       NaN       1.0       NaN  ...           0           0        0   
13589       NaN       NaN       NaN  ...           0           0        0   
13590       NaN       NaN       NaN  ...           0           0        0   
13591       NaN       1.0       NaN  ...           0           0        0   
13592       NaN       1.0       NaN  ...           0           0        0   

           basewgt  adj_mod_basewgt      finalwgt  secu_p  sest  cmintvw  \
0      3410.389399      3869.349602   6448.271112       2     9      NaN   
1      3410.389399      3869.349602   6448.271112       2     9      NaN   
2      7226.301740      8567.549110  12999.542264       2    12      NaN   
3      7226.301740      8567.549110  12999.542264       2    12      NaN   
4      7226.301740      8567.549110  12999.542264       2    12      NaN   
...            ...              ...           ...     ...   ...      ...   
13588  4670.540953      5795.692880   6269.200989       1    78      NaN   
13589  4670.540953      5795.692880   6269.200989       1    78      NaN   
13590  4670.540953      5795.692880   6269.200989       1    78      NaN   
13591  4670.540953      5795.692880   6269.200989       1    78      NaN   
13592  4670.540953      5795.692880   6269.200989       1    78      NaN   

       totalwgt_lb  
0           8.8125  
1           7.8750  
2           9.1250  
3           7.0000  
4           6.1875  
...            ...  
13588       6.1875  
13589          NaN  
13590          NaN  
13591       7.5000  
13592       7.5000  

[13593 rows x 244 columns]

In [10]:
df.columns

Index(['caseid', 'pregordr', 'howpreg_n', 'howpreg_p', 'moscurrp', 'nowprgdk',
       'pregend1', 'pregend2', 'nbrnaliv', 'multbrth',
       ...
       'laborfor_i', 'religion_i', 'metro_i', 'basewgt', 'adj_mod_basewgt',
       'finalwgt', 'secu_p', 'sest', 'cmintvw', 'totalwgt_lb'],
      dtype='object', length=244)

## Create DataFrame of live births

In [14]:
live = df[df.outcome == 1]
live

caseid  pregordr  howpreg_n  howpreg_p  moscurrp  nowprgdk  pregend1  \
0           1         1        NaN        NaN       NaN       NaN       6.0   
1           1         2        NaN        NaN       NaN       NaN       6.0   
2           2         1        NaN        NaN       NaN       NaN       5.0   
3           2         2        NaN        NaN       NaN       NaN       6.0   
4           2         3        NaN        NaN       NaN       NaN       6.0   
...       ...       ...        ...        ...       ...       ...       ...   
13581   12568         2        NaN        NaN       NaN       NaN       5.0   
13584   12569         2        NaN        NaN       NaN       NaN       6.0   
13588   12571         1        NaN        NaN       NaN       NaN       6.0   
13591   12571         4        NaN        NaN       NaN       NaN       6.0   
13592   12571         5        NaN        NaN       NaN       NaN       6.0   

       pregend2  nbrnaliv  multbrth  ...  laborfor_i  religion_i  metro_i  \
0           NaN       1.0       NaN  ...           0           0        0   
1           NaN       1.0       NaN  ...           0           0        0   
2           NaN       3.0       5.0  ...           0           0        0   
3           NaN       1.0       NaN  ...           0           0        0   
4           NaN       1.0       NaN  ...           0           0        0   
...         ...       ...       ...  ...         ...         ...      ...   
13581       NaN       1.0       NaN  ...           0           0        0   
13584       NaN       1.0       NaN  ...           0           0        0   
13588       NaN       1.0       NaN  ...           0           0        0   
13591       NaN       1.0       NaN  ...           0           0        0   
13592       NaN       1.0       NaN  ...           0           0        0   

           basewgt  adj_mod_basewgt      finalwgt  secu_p  sest  cmintvw  \
0      3410.389399      3869.349602   6448.271112       2     9      NaN   
1      3410.389399      3869.349602   6448.271112       2     9      NaN   
2      7226.301740      8567.549110  12999.542264       2    12      NaN   
3      7226.301740      8567.549110  12999.542264       2    12      NaN   
4      7226.301740      8567.549110  12999.542264       2    12      NaN   
...            ...              ...           ...     ...   ...      ...   
13581  2734.687353      4258.980140   7772.212858       2    28      NaN   
13584  2580.967613      2925.167116   5075.164946       2    61      NaN   
13588  4670.540953      5795.692880   6269.200989       1    78      NaN   
13591  4670.540953      5795.692880   6269.200989       1    78      NaN   
13592  4670.540953      5795.692880   6269.200989       1    78      NaN   

       totalwgt_lb  
0           8.8125  
1           7.8750  
2           9.1250  
3           7.0000  
4           6.1875  
...            ...  
13581       6.3750  
13584       6.3750  
13588       6.1875  
13591       7.5000  
13592       7.5000  

[9148 rows x 244 columns]

## Create DataFrames of first and other births, plot them

In [15]:
firsts = live[live.pregordr == 1]
others = live[live.pregordr != 1]

In [35]:
first_hist = thinkstats2.Hist(firsts.prglngth, label='first')
other_hist = thinkstats2.Hist(others.prglngth, label='other')

# width = 0.45
# thinkplot.PrePlot(2)
# thinkplot.Hist(first_hist, align='right', width=width)
# thinkplot.Hist(other_hist, align='left', width=width)
# thinkplot.Show(xlabel='weeks', ylabel='frequency', xlim=[27, 46])

In [22]:
firsts.totalwgt_lb

0        8.8125
2        9.1250
5        8.5625
8        7.5625
10       7.8125
          ...  
13569    5.8125
13573    6.5625
13576    6.4375
13578    6.0000
13588    6.1875
Name: totalwgt_lb, Length: 3368, dtype: float64

## Calculate Cohen's *d* for totalwgt_lb

An effect size is a summary statistic intended to describe (wait for it) the size of an effect. For example, to describe the difference between two groups, one obvious choice is the difference in the means.

Cohen’s d is a statistic intended to do that; it is defined

###  *d* = (x_1 − x_2) / s
  
where __x_1__ and __x_2__ are the means of the groups and __s__ is the “pooled standard deviation”.

In [27]:
from math import sqrt
def cohens_d(group1, group2):
    mean1 = group1.mean()
    mean2 = group2.mean()
    
    var1 = group1.var()
    var2 = group2.var()
    n1, n2 = len(group1), len(group2)
    
    pooled_var = (n1*var1 + n2* var2) / (n1 + n2)
    d = (mean1 - mean2) / sqrt(pooled_var)
    
    return d

## Cohen's *d* for totalwgt_lb is: 

In [33]:
totalwgt_lb_d = cohens_d(firsts.totalwgt_lb, others.totalwgt_lb)

totalwgt_lb_d
'%s' % float('%.2g' % totalwgt_lb_d)

'-0.069'

## And for pregnancy length prglngth:

In [34]:
prglngth_d = cohens_d(firsts.prglngth, others.prglngth)

prglngth_d
'%s' % float('%.2g' % prglngth_d)

'0.014'

## Cohen's *d* is larger in magnitude for total weight than for pregnancy length, when comparing first babies to later babies (-0.069 vs 0.014); however, whether such an effect is statistically significant is unknown. (It probably isn't, though.)